### ISSUE TITLE
e2e test keeps trying the wrong server

### ISSUE THREAD

**Issue by mml** [kubernetes/kubernetes#31285]
I could not figure out why my e2e test kept trying to contact the wrong server, so I added debugging output to enumerate all the Cluster entries in the `clientcmdapi.Config`...

```
Aug 23 12:11:37.854: INFO: >>> kubeConfig: /usr/local/google/home/mml/.kube/config

Aug 23 12:11:37.862: INFO: >>> Clusters["gke_mml-fed_asia-east1-c_ae1c"].Server = "https://104.199.190.143"
Aug 23 12:11:37.862: INFO: >>> Clusters["gke_mml-fed_us-central1-a_uc1a"].Server = "https://23.251.150.221"
Aug 23 12:11:37.862: INFO: >>> Clusters["local"].Server = "http://127.0.0.1:8080"
Aug 23 12:11:37.862: INFO: >>> Clusters["mml-e2e_e2e"].Server = "https://104.197.209.171"
Aug 23 12:11:37.862: INFO: >>> Clusters["federation-cluster"].Server = "https://104.155.185.175:443"
Aug 23 12:11:37.862: INFO: >>> Clusters["federation-e2e-gce-us-central1-b"].Server = "https://104.197.213.50"
Aug 23 12:11:37.862: INFO: >>> Clusters["federation-e2e-gce-us-central1-c"].Server = "https://104.155.185.116"
Aug 23 12:11:37.863: INFO: Waiting up to 10m0s for all pods (need at least 0) in namespace 'kube-system' to be running and ready
Aug 23 12:12:07.864: INFO: Error getting replication controllers in namespace 'kube-system': Get https://104.198.199.192/api/v1/namespaces/kube-system/replicationcontrollers: dial tcp 104.198.199.192:443: i/o timeout
Aug 23 12:12:07.864: INFO: Error getting pods in namespace "kube-system": Get https://104.198.199.192/api/v1/namespaces/kube-system/pods?labelSelector=name%3De2e-image-puller: dial tcp 104.198.199.192:443: i/o timeout
```

Where the heck is `104.198.199.192` coming from?  Still looking.

@kubernetes/sig-api-machinery

**Comment by mml** [comment]
Huh.  This is due in part to the bug I previously fixed in #29038, which looks like it got reverted.

If I have `FEDERATION=false`, `e2e.go` enters the federation path.  It sets `OVERRIDE_CONTEXT` env var.

```
2016/08/23 12:42:59 e2e.go:451: Running: Federated Ginkgo tests
Setting zone vars to: federation-e2e-gce-us-central1-f
Setting up for KUBERNETES_PROVIDER="gce".
Project: mml-e2e
Zone: us-central1-f
```

But I don't have a context named `federation-e2e-gce-us-central1-f`, so where is it getting that address?

**Comment by mml** [comment]
OK, if we run `hack/federated-ginkgo-e2e.sh`, then we set `KUBE_GCE_INSTANCE_PREFIX=mml-us-central1-f`.  In `cluster/gce/config-default.sh`, we then use that to set `INSTANCE_PREFIX`.  And we set `MASTER_NAME` to `${INSTANCE_PREFIX}-master`.

Then, deep in the bowels of `cluster/gce/util.sh`, in `detect-master`, we try to find an address for the master by guessing the name of its address: `${MASTER_NAME}-ip`.

```console
% gcloud compute addresses describe mml-us-central1-f-master-ip --project mml-e2e --region us-central1 -q --format='value(address)'
104.198.199.192
```

That IP address is still around, probably because I gracelessly ripped down a cluster long ago.  In most cases if this code ever gets run, I have actually brought up a federated cluster and we use that one instead.

The communication via environment variables set all over the place is a known bug that I'm not going to do anything about now.  I can supply one small fix that stops running this code when `FEDERATION != true` and another one to improve our ability to debug this.
